src/EditorFeatures/CSharpTest2/Recommendations/ForEachKeywordRecommenderTests.cs

```diff
@@ -41,123 +56,139 @@ public async Task TestAfterGlobalVariableDeclaration_Interactive()
-        public async Task TestNotAfterForEach1()
+        [Theory, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        [CombinatorialData]
+        public async Task TestNotAfterForEach1(bool topLevelStatement)
         {
             await VerifyAbsenceAsync(AddInsideMethod(
-@"foreach $$"));
+@"foreach $$", topLevelStatement: topLevelStatement), options: CSharp9ParseOptions);
         }
 
-        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
-        public async Task TestNotAfterForEach2()
+        [Theory, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        [CombinatorialData]
+        public async Task TestNotAfterForEach2(bool topLevelStatement)
         {
             await VerifyAbsenceAsync(AddInsideMethod(
-@"foreach ($$"));
+@"foreach ($$", topLevelStatement: topLevelStatement), options: CSharp9ParseOptions);
         }
 
-        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
-        public async Task TestNotAfterForEach3()
+        [Theory, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        [CombinatorialData]
+        public async Task TestNotAfterForEach3(bool topLevelStatement)
         {
             await VerifyAbsenceAsync(AddInsideMethod(
-@"foreach (var $$"));
+@"foreach (var $$", topLevelStatement: topLevelStatement), options: CSharp9ParseOptions);
         }
 
-        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
-        public async Task TestNotAfterForEach4()
+        [Theory, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        [CombinatorialData]
+        public async Task TestNotAfterForEach4(bool topLevelStatement)
         {
             await VerifyAbsenceAsync(AddInsideMethod(
-@"foreach (var v $$"));
+@"foreach (var v $$", topLevelStatement: topLevelStatement), options: CSharp9ParseOptions);
         }
 
-        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
-        public async Task TestNotAfterForEach5()
+        [Theory, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        [CombinatorialData]
+        public async Task TestNotAfterForEach5(bool topLevelStatement)
         {
             await VerifyAbsenceAsync(AddInsideMethod(
-@"foreach (var v in $$"));
+@"foreach (var v in $$", topLevelStatement: topLevelStatement), options: CSharp9ParseOptions);
         }
 
-        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
-        public async Task TestNotAfterForEach6()
+        [Theory, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        [CombinatorialData]
+        public async Task TestNotAfterForEach6(bool topLevelStatement)
         {
             await VerifyAbsenceAsync(AddInsideMethod(
-@"foreach (var v in c $$"));
+@"foreach (var v in c $$", topLevelStatement: topLevelStatement), options: CSharp9ParseOptions);
         }
 
         [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
```

Comment:
do we have testst that validate that if a using-directive is *after* the caret that we don't offer lots of these keywords before it?  i.e. we already filter out things like `namespace/class/struct` before using directives.  we'd want to do that for the rest of the top-level keywords.

Reply:
Some keyword recommenders tested this for script code already. For the test classes I updated, I preserved these tests for top-level statements.